scripts: coccicheck: Improve continuous run operation

himanshujha199640 opened the pull request:
The current implementation of continuous run operation using
command `./scripts/coccicheck` i.e., without specifying any options,
`coccicheck` default runs in `report` mode with all available
coccinelle scripts present at `scripts/coccinelle/`.

Not all scripts have report mode implemented in them, which
leads to failure of coccicheck.

With this new implementation we choose whatever available mode
is present in coccinelle script and pass it to MODE variable
without stopping continuous coverage.

And perhaps if there are plans to add `coccicheck` as a sanity
checker in future to the Zephyr automated CI, then certainly we
want the warnings/errors produced by scripts to be less verbose
to the users.

Therefore, in this new implementation we prioritise the modes as:

1. report
2. context
3. patch
and lastly falling to
4. org

Lastly, in order to differentiate between outputs of various
coccinelle scripts being run, `x------x` separator has been used
to make reports mode readable.

Signed-off-by: Himanshu Jha <himanshujha199640@gmail.com>

Cc @JuliaLawall @nashif 

Sample run with some statistics:
```
himanshu@himanshu-Vostro-3559:~/zephyr$ time ./scripts/coccicheck --verbose=1
You have not explicitly specified the mode to use. Using default "report" mode.
Available modes are the following: patch, report, context, org
You can specify the mode with "./scripts/coccicheck --mode=<mode>"
Note however that some modes are not implemented by some semantic patches.

Please check for false positives in the output before submitting a patch.
When using "patch" mode, carefully review the patch before submitting it.

Processing array_size.cocci
with option(s) " --no-includes --include-headers"

Message example to submit a patch:
 Use ARRAY_SIZE instead of dividing sizeof array with sizeof an element

 The semantic patch that makes this report is available
 in scripts/coccinelle/array_size.cocci.

 More information about semantic patching is available at
 http://coccinelle.lip6.fr/

Semantic patch information:
 This makes an effort to find cases where ARRAY_SIZE can be used such as
 where there is a division of sizeof the array by the sizeof its first
 element or by any indexed element or the element type. It replaces the
 division of the two sizeofs by ARRAY_SIZE.

Running (4 in parallel): /usr/local/bin/spatch -D report --no-show-diff --very-quiet --cocci-file /home/himanshu/zephyr/scripts/coccinelle/array_size.cocci --no-includes --include-headers --dir /home/himanshu/zephyr --jobs 4 --chunksize 1
X-------------------------------------------------------------------------X

Processing deref_null.cocci
with option(s) ""

Message example to submit a patch:

 A variable is dereferenced under a NULL test.
 Even though it is known to be NULL.

 The semantic patch that makes this report is available
 in scripts/coccinelle/deref_null.cocci.

 More information about semantic patching is available at
 http://coccinelle.lip6.fr/

Running (4 in parallel): /usr/local/bin/spatch -D report --no-show-diff --very-quiet --cocci-file /home/himanshu/zephyr/scripts/coccinelle/deref_null.cocci --dir /home/himanshu/zephyr --jobs 4 --chunksize 1
X-------------------------------------------------------------------------X

Invalid mode "report" supplied!
Available modes for "ignore_return.cocci" are: patch
Using random availble mode: "patch"

Processing ignore_return.cocci
with option(s) ""

Message example to submit a patch:
 Cast void to memset to ignore its return value

 The semantic patch that makes this change is available
 in scripts/coccinelle/ignore_return.cocci.

 More information about semantic patching is available at
 http://coccinelle.lip6.fr/

Semantic patch information:
 The return of memset is never checked and therefore cast
 it to void to explicitly ignore while adhering to MISRA-C.

Running (4 in parallel): /usr/local/bin/spatch -D patch --no-show-diff --very-quiet --cocci-file /home/himanshu/zephyr/scripts/coccinelle/ignore_return.cocci --dir /home/himanshu/zephyr --jobs 4 --chunksize 1
X-------------------------------------------------------------------------X

Invalid mode "report" supplied!
Available modes for "irq_lock.cocci" are: patch
Using random availble mode: "patch"

Processing irq_lock.cocci
with option(s) ""

Message example to submit a patch:
 Use unsigned int as the return value for irq_lock()

 The semantic patch that makes this change is available
 in scripts/coccinelle/irq_lock.cocci.

 More information about semantic patching is available at
 http://coccinelle.lip6.fr/

Running (4 in parallel): /usr/local/bin/spatch -D patch --no-show-diff --very-quiet --cocci-file /home/himanshu/zephyr/scripts/coccinelle/irq_lock.cocci --dir /home/himanshu/zephyr --jobs 4 --chunksize 1
X-------------------------------------------------------------------------X

Processing mini_lock.cocci
with option(s) ""

Message example to submit a patch:
 Find missing unlocks.  This semantic match considers the specific case
 where the unlock is missing from an if branch, and there is a lock
 before the if and an unlock after the if.  False positives are due to
 cases where the if branch represents a case where the function is
 supposed to exit with the lock held, or where there is some preceding
 function call that releases the lock.

 The semantic patch that makes this report is available
 in scripts/coccinelle/mini_lock.cocci.

 More information about semantic patching is available at
 http://coccinelle.lip6.fr/

Running (4 in parallel): /usr/local/bin/spatch -D report --no-show-diff --very-quiet --cocci-file /home/himanshu/zephyr/scripts/coccinelle/mini_lock.cocci --dir /home/himanshu/zephyr --jobs 4 --chunksize 1
/home/himanshu/zephyr/subsys/bluetooth/host/rfcomm.c:519:2-8: preceding lock on line 516
X-------------------------------------------------------------------------X

Processing noderef.cocci
with option(s) " --no-includes --include-headers"

Message example to submit a patch:
 sizeof when applied to a pointer typed expression gives the size of
 the pointer

 The semantic patch that makes this report is available
 in scripts/coccinelle/noderef.cocci.

 More information about semantic patching is available at
 http://coccinelle.lip6.fr/

Running (4 in parallel): /usr/local/bin/spatch -D report --no-show-diff --very-quiet --cocci-file /home/himanshu/zephyr/scripts/coccinelle/noderef.cocci --no-includes --include-headers --dir /home/himanshu/zephyr --jobs 4 --chunksize 1
/home/himanshu/zephyr/subsys/bluetooth/controller/util/mem.c:79:23-29: ERROR: application of sizeof to pointer
X-------------------------------------------------------------------------X

Processing semicolon.cocci
with option(s) " --no-includes --include-headers"

Message example to submit a patch:

 Remove unneeded semicolon.

 The semantic patch that makes this report is available
 in scripts/coccinelle/semicolon.cocci.

 More information about semantic patching is available at
 http://coccinelle.lip6.fr/

Running (4 in parallel): /usr/local/bin/spatch -D report --no-show-diff --very-quiet --cocci-file /home/himanshu/zephyr/scripts/coccinelle/semicolon.cocci --no-includes --include-headers --dir /home/himanshu/zephyr --jobs 4 --chunksize 1
/home/himanshu/zephyr/lib/libc/minimal/source/string/string.c:253:3-4: Unneeded semicolon
/home/himanshu/zephyr/lib/libc/minimal/source/string/string.c:299:2-3: Unneeded semicolon
/home/himanshu/zephyr/samples/boards/bbc_microbit/pong/src/main.c:215:2-3: Unneeded semicolon
/home/himanshu/zephyr/subsys/bluetooth/host/hci_core.c:2019:3-4: Unneeded semicolon
/home/himanshu/zephyr/subsys/bluetooth/host/hci_core.c:2240:3-4: Unneeded semicolon
/home/himanshu/zephyr/subsys/bluetooth/shell/bredr.c:118:3-4: Unneeded semicolon
/home/himanshu/zephyr/tests/drivers/i2c/i2c_slave_api/common/i2c_virtual.c:203:2-3: Unneeded semicolon
/home/himanshu/zephyr/samples/net/wpan_serial/src/main.c:634:2-3: Unneeded semicolon
/home/himanshu/zephyr/samples/boards/up_squared/gpio_counter/src/main.c:127:2-3: Unneeded semicolon
/home/himanshu/zephyr/subsys/bluetooth/host/l2cap_br.c:620:2-3: Unneeded semicolon
/home/himanshu/zephyr/tests/drivers/watchdog/wdt_basic_api/src/test_wdt.c:201:2-3: Unneeded semicolon
/home/himanshu/zephyr/tests/drivers/watchdog/wdt_basic_api/src/test_wdt.c:262:2-3: Unneeded semicolon
/home/himanshu/zephyr/tests/drivers/watchdog/wdt_basic_api/src/test_wdt.c:154:2-3: Unneeded semicolon
/home/himanshu/zephyr/samples/drivers/watchdog/src/main.c:74:2-3: Unneeded semicolon
/home/himanshu/zephyr/subsys/net/ip/icmpv6.c:157:42-43: Unneeded semicolon
/home/himanshu/zephyr/subsys/net/ip/icmpv6.c:191:45-46: Unneeded semicolon
/home/himanshu/zephyr/subsys/bluetooth/host/smp.c:3785:2-3: Unneeded semicolon
/home/himanshu/zephyr/subsys/bluetooth/host/smp.c:3791:2-3: Unneeded semicolon
X-------------------------------------------------------------------------X

Processing unsigned_lesser_than_zero.cocci
with option(s) ""

Message example to submit a patch:
 Unsigned expressions cannot be lesser than zero. Presence of
 comparisons 'unsigned (<|<=) 0' often indicates a bug,
 usually wrong type of variable.

 The semantic patch that makes this report is available
 in scripts/coccinelle/unsigned_lesser_than_zero.cocci.

 More information about semantic patching is available at
 http://coccinelle.lip6.fr/

Running (4 in parallel): /usr/local/bin/spatch -D report --no-show-diff --very-quiet --cocci-file /home/himanshu/zephyr/scripts/coccinelle/unsigned_lesser_than_zero.cocci --dir /home/himanshu/zephyr --jobs 4 --chunksize 1
/home/himanshu/zephyr/drivers/spi/spi_sam0.c:231:5-8: WARNING: Unsigned expression compared with zero.
/home/himanshu/zephyr/subsys/net/lib/lwm2m/lwm2m_obj_device.c:172:5-16: WARNING: Unsigned expression compared with zero.
/home/himanshu/zephyr/subsys/net/lib/mqtt/mqtt_pkt.c:638:5-10: WARNING: Unsigned expression compared with zero.
/home/himanshu/zephyr/subsys/net/lib/mqtt/mqtt_pkt.c:793:5-13: WARNING: Unsigned expression compared with zero.
/home/himanshu/zephyr/lib/posix/mqueue.c:71:45-53: WARNING: Unsigned expression compared with zero.
/home/himanshu/zephyr/lib/posix/mqueue.c:72:10-18: WARNING: Unsigned expression compared with zero.
/home/himanshu/zephyr/drivers/flash/soc_flash_nios2_qspi.c:90:7-23: WARNING: Unsigned expression compared with zero.
/home/himanshu/zephyr/drivers/flash/soc_flash_nios2_qspi.c:283:6-22: WARNING: Unsigned expression compared with zero.
/home/himanshu/zephyr/drivers/spi/spi_sam.c:223:5-8: WARNING: Unsigned expression compared with zero.
X-------------------------------------------------------------------------X

Invalid mode "report" supplied!
Available modes for "unsigned_shift.cocci" are: patch
Using random availble mode: "patch"

Processing unsigned_shift.cocci
with option(s) ""

Message example to submit a patch:
 Use BIT() helper macro instead of hardcoding using bitshifting

 The semantic patch that makes this change is available
 in scripts/coccinelle/unsigned_shift.cocci.

 More information about semantic patching is available at
 http://coccinelle.lip6.fr/

Running (4 in parallel): /usr/local/bin/spatch -D patch --no-show-diff --very-quiet --cocci-file /home/himanshu/zephyr/scripts/coccinelle/unsigned_shift.cocci --dir /home/himanshu/zephyr --jobs 4 --chunksize 1
X-------------------------------------------------------------------------X


real	9m40.874s
user	36m58.083s
sys	0m15.343s

himanshu@himanshu-Vostro-3559:~/zephyr$ find scripts/coccinelle/ -name "*.cocci" | wc -l
9

himanshu@himanshu-Vostro-3559:~/zephyr$ lscpu 
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
CPU(s):              4
On-line CPU(s) list: 0-3
Thread(s) per core:  2
Core(s) per socket:  2
Socket(s):           1
NUMA node(s):        1
Vendor ID:           GenuineIntel
CPU family:          6
Model:               78
Model name:          Intel(R) Core(TM) i5-6200U CPU @ 2.30GHz
Stepping:            3
CPU MHz:             1191.070
CPU max MHz:         2800.0000
CPU min MHz:         400.0000
BogoMIPS:            4800.00
Virtualization:      VT-x
L1d cache:           32K
L1i cache:           32K
L2 cache:            256K
L3 cache:            3072K
NUMA node0 CPU(s):   0-3
```

himanshujha199640 committed:
scripts: coccicheck: Improve continuous run operation

The current implementation of continuous run operation using
command `./scripts/coccicheck` i.e., without specifying any options,
`coccicheck` default runs in `report` mode with all available
coccinelle scripts present at `scripts/coccinelle/`.

Not all scripts have report mode implemented in them, which
leads to failure of coccicheck.

With this new implementation we choose whatever available mode
is present in coccinelle script and pass it to MODE variable
without stopping continuous coverage.

And perhaps if there are plans to add `coccicheck` as a sanity
checker in future to the Zephyr automated CI, then certainly we
want the warnings/errors produced by scripts to be less verbose
to the users.

Therefore, in this new implementation we prioritise the modes as:

1. report
2. context
3. patch
and lastly falling to
4. org

Lastly, in order to differentiate between outputs of various
coccinelle scripts being run, `x------x` separator has been used
to make reports mode readable.

Signed-off-by: Himanshu Jha <himanshujha199640@gmail.com>

JuliaLawall commented:
On Mon, 19 Nov 2018, Himanshu Jha wrote:

>
> The current implementation of continuous run operation using
> command ./scripts/coccicheck i.e., without specifying any options,
> coccicheck default runs in report mode with all available
> coccinelle scripts present at scripts/coccinelle/.
>
> Not all scripts have report mode implemented in them, which
> leads to failure of coccicheck.

While it's fine to make coccicheck more robust, would it also be possible
to always define the report mode?

I'm not sure that the Xs at the beginning and end of the separator line
add much.

julia

>
> With this new implementation we choose whatever available mode
> is present in coccinelle script and pass it to MODE variable
> without stopping continuous coverage.
>
> And perhaps if there are plans to add coccicheck as a sanity
> checker in future to the Zephyr automated CI, then certainly we
> want the warnings/errors produced by scripts to be less verbose
> to the users.
>
> Therefore, in this new implementation we prioritise the modes as:
>
>  1. report
>  2. context
>  3. patch
>     and lastly falling to
>  4. org
>
> Lastly, in order to differentiate between outputs of various
> coccinelle scripts being run, x------x separator has been used
> to make reports mode readable.
>
> Signed-off-by: Himanshu Jha himanshujha199640@gmail.com
>
> Cc @JuliaLawall @nashif
>
> Sample run with some statistics:
>
> himanshu@himanshu-Vostro-3559:~/zephyr$ time ./scripts/coccicheck --verbose=
> 1
> You have not explicitly specified the mode to use. Using default "report" mo
> de.
> Available modes are the following: patch, report, context, org
> You can specify the mode with "./scripts/coccicheck --mode=<mode>"
> Note however that some modes are not implemented by some semantic patches.
>
> Please check for false positives in the output before submitting a patch.
> When using "patch" mode, carefully review the patch before submitting it.
>
> Processing array_size.cocci
> with option(s) " --no-includes --include-headers"
>
> Message example to submit a patch:
>  Use ARRAY_SIZE instead of dividing sizeof array with sizeof an element
>
>  The semantic patch that makes this report is available
>  in scripts/coccinelle/array_size.cocci.
>
>  More information about semantic patching is available at
>  http://coccinelle.lip6.fr/
>
> Semantic patch information:
>  This makes an effort to find cases where ARRAY_SIZE can be used such as
>  where there is a division of sizeof the array by the sizeof its first
>  element or by any indexed element or the element type. It replaces the
>  division of the two sizeofs by ARRAY_SIZE.
>
> Running (4 in parallel): /usr/local/bin/spatch -D report --no-show-diff --ve
> ry-quiet --cocci-file /home/himanshu/zephyr/scripts/coccinelle/array_size.co
> cci --no-includes --include-headers --dir /home/himanshu/zephyr --jobs 4 --c
> hunksize 1
> X-------------------------------------------------------------------------X
>
> Processing deref_null.cocci
> with option(s) ""
>
> Message example to submit a patch:
>
>  A variable is dereferenced under a NULL test.
>  Even though it is known to be NULL.
>
>  The semantic patch that makes this report is available
>  in scripts/coccinelle/deref_null.cocci.
>
>  More information about semantic patching is available at
>  http://coccinelle.lip6.fr/
>
> Running (4 in parallel): /usr/local/bin/spatch -D report --no-show-diff --ve
> ry-quiet --cocci-file /home/himanshu/zephyr/scripts/coccinelle/deref_null.co
> cci --dir /home/himanshu/zephyr --jobs 4 --chunksize 1
> X-------------------------------------------------------------------------X
>
> Invalid mode "report" supplied!
> Available modes for "ignore_return.cocci" are: patch
> Using random availble mode: "patch"
>
> Processing ignore_return.cocci
> with option(s) ""
>
> Message example to submit a patch:
>  Cast void to memset to ignore its return value
>
>  The semantic patch that makes this change is available
>  in scripts/coccinelle/ignore_return.cocci.
>
>  More information about semantic patching is available at
>  http://coccinelle.lip6.fr/
>
> Semantic patch information:
>  The return of memset is never checked and therefore cast
>  it to void to explicitly ignore while adhering to MISRA-C.
>
> Running (4 in parallel): /usr/local/bin/spatch -D patch --no-show-diff --ver
> y-quiet --cocci-file /home/himanshu/zephyr/scripts/coccinelle/ignore_return.
> cocci --dir /home/himanshu/zephyr --jobs 4 --chunksize 1
> X-------------------------------------------------------------------------X
>
> Invalid mode "report" supplied!
> Available modes for "irq_lock.cocci" are: patch
> Using random availble mode: "patch"
>
> Processing irq_lock.cocci
> with option(s) ""
>
> Message example to submit a patch:
>  Use unsigned int as the return value for irq_lock()
>
>  The semantic patch that makes this change is available
>  in scripts/coccinelle/irq_lock.cocci.
>
>  More information about semantic patching is available at
>  http://coccinelle.lip6.fr/
>
> Running (4 in parallel): /usr/local/bin/spatch -D patch --no-show-diff --ver
> y-quiet --cocci-file /home/himanshu/zephyr/scripts/coccinelle/irq_lock.cocci
>  --dir /home/himanshu/zephyr --jobs 4 --chunksize 1
> X-------------------------------------------------------------------------X
>
> Processing mini_lock.cocci
> with option(s) ""
>
> Message example to submit a patch:
>  Find missing unlocks.  This semantic match considers the specific case
>  where the unlock is missing from an if branch, and there is a lock
>  before the if and an unlock after the if.  False positives are due to
>  cases where the if branch represents a case where the function is
>  supposed to exit with the lock held, or where there is some preceding
>  function call that releases the lock.
>
>  The semantic patch that makes this report is available
>  in scripts/coccinelle/mini_lock.cocci.
>
>  More information about semantic patching is available at
>  http://coccinelle.lip6.fr/
>
> Running (4 in parallel): /usr/local/bin/spatch -D report --no-show-diff --ve
> ry-quiet --cocci-file /home/himanshu/zephyr/scripts/coccinelle/mini_lock.coc
> ci --dir /home/himanshu/zephyr --jobs 4 --chunksize 1
> /home/himanshu/zephyr/subsys/bluetooth/host/rfcomm.c:519:2-8: preceding lock
>  on line 516
> X-------------------------------------------------------------------------X
>
> Processing noderef.cocci
> with option(s) " --no-includes --include-headers"
>
> Message example to submit a patch:
>  sizeof when applied to a pointer typed expression gives the size of
>  the pointer
>
>  The semantic patch that makes this report is available
>  in scripts/coccinelle/noderef.cocci.
>
>  More information about semantic patching is available at
>  http://coccinelle.lip6.fr/
>
> Running (4 in parallel): /usr/local/bin/spatch -D report --no-show-diff --ve
> ry-quiet --cocci-file /home/himanshu/zephyr/scripts/coccinelle/noderef.cocci
>  --no-includes --include-headers --dir /home/himanshu/zephyr --jobs 4 --chun
> ksize 1
> /home/himanshu/zephyr/subsys/bluetooth/controller/util/mem.c:79:23-29: ERROR
> : application of sizeof to pointer
> X-------------------------------------------------------------------------X
>
> Processing semicolon.cocci
> with option(s) " --no-includes --include-headers"
>
> Message example to submit a patch:
>
>  Remove unneeded semicolon.
>
>  The semantic patch that makes this report is available
>  in scripts/coccinelle/semicolon.cocci.
>
>  More information about semantic patching is available at
>  http://coccinelle.lip6.fr/
>
> Running (4 in parallel): /usr/local/bin/spatch -D report --no-show-diff --ve
> ry-quiet --cocci-file /home/himanshu/zephyr/scripts/coccinelle/semicolon.coc
> ci --no-includes --include-headers --dir /home/himanshu/zephyr --jobs 4 --ch
> unksize 1
> /home/himanshu/zephyr/lib/libc/minimal/source/string/string.c:253:3-4: Unnee
> ded semicolon
> /home/himanshu/zephyr/lib/libc/minimal/source/string/string.c:299:2-3: Unnee
> ded semicolon
> /home/himanshu/zephyr/samples/boards/bbc_microbit/pong/src/main.c:215:2-3: U
> nneeded semicolon
> /home/himanshu/zephyr/subsys/bluetooth/host/hci_core.c:2019:3-4: Unneeded se
> micolon
> /home/himanshu/zephyr/subsys/bluetooth/host/hci_core.c:2240:3-4: Unneeded se
> micolon
> /home/himanshu/zephyr/subsys/bluetooth/shell/bredr.c:118:3-4: Unneeded semic
> olon
> /home/himanshu/zephyr/tests/drivers/i2c/i2c_slave_api/common/i2c_virtual.c:2
> 03:2-3: Unneeded semicolon
> /home/himanshu/zephyr/samples/net/wpan_serial/src/main.c:634:2-3: Unneeded s
> emicolon
> /home/himanshu/zephyr/samples/boards/up_squared/gpio_counter/src/main.c:127:
> 2-3: Unneeded semicolon
> /home/himanshu/zephyr/subsys/bluetooth/host/l2cap_br.c:620:2-3: Unneeded sem
> icolon
> /home/himanshu/zephyr/tests/drivers/watchdog/wdt_basic_api/src/test_wdt.c:20
> 1:2-3: Unneeded semicolon
> /home/himanshu/zephyr/tests/drivers/watchdog/wdt_basic_api/src/test_wdt.c:26
> 2:2-3: Unneeded semicolon
> /home/himanshu/zephyr/tests/drivers/watchdog/wdt_basic_api/src/test_wdt.c:15
> 4:2-3: Unneeded semicolon
> /home/himanshu/zephyr/samples/drivers/watchdog/src/main.c:74:2-3: Unneeded s
> emicolon
> /home/himanshu/zephyr/subsys/net/ip/icmpv6.c:157:42-43: Unneeded semicolon
> /home/himanshu/zephyr/subsys/net/ip/icmpv6.c:191:45-46: Unneeded semicolon
> /home/himanshu/zephyr/subsys/bluetooth/host/smp.c:3785:2-3: Unneeded semicol
> on
> /home/himanshu/zephyr/subsys/bluetooth/host/smp.c:3791:2-3: Unneeded semicol
> on
> X-------------------------------------------------------------------------X
>
> Processing unsigned_lesser_than_zero.cocci
> with option(s) ""
>
> Message example to submit a patch:
>  Unsigned expressions cannot be lesser than zero. Presence of
>  comparisons 'unsigned (<|<=) 0' often indicates a bug,
>  usually wrong type of variable.
>
>  The semantic patch that makes this report is available
>  in scripts/coccinelle/unsigned_lesser_than_zero.cocci.
>
>  More information about semantic patching is available at
>  http://coccinelle.lip6.fr/
>
> Running (4 in parallel): /usr/local/bin/spatch -D report --no-show-diff --ve
> ry-quiet --cocci-file /home/himanshu/zephyr/scripts/coccinelle/unsigned_less
> er_than_zero.cocci --dir /home/himanshu/zephyr --jobs 4 --chunksize 1
> /home/himanshu/zephyr/drivers/spi/spi_sam0.c:231:5-8: WARNING: Unsigned expr
> ession compared with zero.
> /home/himanshu/zephyr/subsys/net/lib/lwm2m/lwm2m_obj_device.c:172:5-16: WARN
> ING: Unsigned expression compared with zero.
> /home/himanshu/zephyr/subsys/net/lib/mqtt/mqtt_pkt.c:638:5-10: WARNING: Unsi
> gned expression compared with zero.
> /home/himanshu/zephyr/subsys/net/lib/mqtt/mqtt_pkt.c:793:5-13: WARNING: Unsi
> gned expression compared with zero.
> /home/himanshu/zephyr/lib/posix/mqueue.c:71:45-53: WARNING: Unsigned express
> ion compared with zero.
> /home/himanshu/zephyr/lib/posix/mqueue.c:72:10-18: WARNING: Unsigned express
> ion compared with zero.
> /home/himanshu/zephyr/drivers/flash/soc_flash_nios2_qspi.c:90:7-23: WARNING:
>  Unsigned expression compared with zero.
> /home/himanshu/zephyr/drivers/flash/soc_flash_nios2_qspi.c:283:6-22: WARNING
> : Unsigned expression compared with zero.
> /home/himanshu/zephyr/drivers/spi/spi_sam.c:223:5-8: WARNING: Unsigned expre
> ssion compared with zero.
> X-------------------------------------------------------------------------X
>
> Invalid mode "report" supplied!
> Available modes for "unsigned_shift.cocci" are: patch
> Using random availble mode: "patch"
>
> Processing unsigned_shift.cocci
> with option(s) ""
>
> Message example to submit a patch:
>  Use BIT() helper macro instead of hardcoding using bitshifting
>
>  The semantic patch that makes this change is available
>  in scripts/coccinelle/unsigned_shift.cocci.
>
>  More information about semantic patching is available at
>  http://coccinelle.lip6.fr/
>
> Running (4 in parallel): /usr/local/bin/spatch -D patch --no-show-diff --ver
> y-quiet --cocci-file /home/himanshu/zephyr/scripts/coccinelle/unsigned_shift
> .cocci --dir /home/himanshu/zephyr --jobs 4 --chunksize 1
> X-------------------------------------------------------------------------X
>
>
> real	9m40.874s
> user	36m58.083s
> sys	0m15.343s
>
> himanshu@himanshu-Vostro-3559:~/zephyr$ find scripts/coccinelle/ -name "*.co
> cci" | wc -l
> 9
>
> himanshu@himanshu-Vostro-3559:~/zephyr$ lscpu
> Architecture:        x86_64
> CPU op-mode(s):      32-bit, 64-bit
> Byte Order:          Little Endian
> CPU(s):              4
> On-line CPU(s) list: 0-3
> Thread(s) per core:  2
> Core(s) per socket:  2
> Socket(s):           1
> NUMA node(s):        1
> Vendor ID:           GenuineIntel
> CPU family:          6
> Model:               78
> Model name:          Intel(R) Core(TM) i5-6200U CPU @ 2.30GHz
> Stepping:            3
> CPU MHz:             1191.070
> CPU max MHz:         2800.0000
> CPU min MHz:         400.0000
> BogoMIPS:            4800.00
> Virtualization:      VT-x
> L1d cache:           32K
> L1i cache:           32K
> L2 cache:            256K
> L3 cache:            3072K
> NUMA node0 CPU(s):   0-3
>
> ____________________________________________________________________________
>
>     You can view, comment on, or merge this pull request online at:
>
>   https://github.com/zephyrproject-rtos/zephyr/pull/11506
>
>     Commit Summary
>
>  *  scripts: coccicheck: Improve continuous run operation
>
>     File Changes
>
>  *  M scripts/coccicheck (20)
>
>     Patch Links:
>
>  *  https://github.com/zephyrproject-rtos/zephyr/pull/11506.patch
>  *  https://github.com/zephyrproject-rtos/zephyr/pull/11506.diff
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the
> thread.[AAesmvp5Mu-Af1IRRWp7G-F7xt5jUG18ks5uwt3ggaJpZM4YpeMg.gif]
>
>
>

himanshujha199640 commented:
> While it's fine to make coccicheck more robust, would it also be possible to always define the report mode? I'm not sure that the Xs at the beginning and end of the separator line add much.

"define" as in all scripts should posses at the very least `report` mode ?

I will add that as a necessary point in `coccinelle.rst` in "Proposing new semantic patches" section 
to always add `report` mode.

Sure, will remove the Xs as well in the next revision!

JuliaLawall commented:
On Mon, 19 Nov 2018, Himanshu Jha wrote:

>
>       While it's fine to make coccicheck more robust, would it also be
>       possible to always define the report mode? I'm not sure that the
>       Xs at the beginning and end of the separator line add much.
>
> "define" as in all scripts should posses at the very least report mode ?

Yes, this is what is done for the Linux kernel at least.

>
> I will add that as a necessary point in coccinelle.rst in "Proposing new
> semantic patches" section
> to always add report mode.
>
> Sure, will remove the Xs as well in the next revision!

Thanks!

julia

> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the
> thread.[AAesmtFMLfeh3CmlbdqLs86OJcjx_VgKks5uwuVWgaJpZM4YpeMg.gif]
>
>
>

ceolin commented on scripts/coccicheck:
```diff
@@ -201,6 +212,7 @@ coccinelle () {
 	run_cmd_parmap $SPATCH -D $MODE   $FLAGS --cocci-file $COCCI $OPT $OPTIONS || exit 1
     fi
 
+    MODE=report
```
what does it doing here ?

himanshujha199640 replied:
The covers the case when suppose a script has no report mode and instead patch mode(available)
is run. This is fine behavior.
But the problem is that next upcoming script to be run also uses the previous propagated
used mode which was patch mode and not report mode.
So, the rule is to keep report mode as the top priority and be less verbose in warnings/reports. Therefore to fulfil such a condition I overwrite MODE variable with report, so that upcoming
script to be used runs in report mode(if available).

Now, as @JuliaLawall suggested it would be better to support at the very least report mode in
all coccinelle scripts in the current and future scripts. That would be really helpful in case we
add coccicheck in the automated CI.

himanshujha199640 committed:
doc: application: coccinelle: Mention report mode as a necessary rule

The newly proposed script must have report mode implemented
in them since they are less verbose warnings and helpful
in automated CI when throwing warnings to users.

Also, fixup mailing list links to follow .rst rules for
documentation using bullets while rectifying unwanted bold text.

Signed-off-by: Himanshu Jha <himanshujha199640@gmail.com>

dbkinder commented on doc/application/coccinelle.rst:
```diff
@@ -513,8 +515,8 @@ Coccinelle Mailing List
 
 Interested developers can subscribe coccinelle mailing list via:
-        https://systeme.lip6.fr/mailman/listinfo/cocci
+* https://systeme.lip6.fr/mailman/listinfo/cocci
```
a blank line is needed before the first item in a bullet list

Also, on the line above, change to ``subscribe to the coccinelle mailing list``

dbkinder commented on doc/application/coccinelle.rst:
```diff
@@ -513,8 +515,8 @@ Coccinelle Mailing List
 Archives:
-        https://lore.kernel.org/cocci/
-        https://systeme.lip6.fr/pipermail/cocci/
+* https://lore.kernel.org/cocci/
```
a blank line is needed before the first item in a bullet list

himanshujha199640 commented:
@dbkinder done!
Thanks for the review.

dbkinder reviewed (approved):
+1 for doc changes
